### DISCUSSION TITLE
API Discussion

### DISCUSSION THREAD

**Discussion by jbtronics** [Part-DB/Part-DB-server#281] [General]
Many users have asked for an API to interface with Part-DB with their own code / software.

What are your applications for that / what do you want to do with the API?

And what features do you see as important for that? What data do you need to access/edit?

**Comment by virtadpt** [comment]
I want to write a bot with a conversational interface for searching my inventory.  "Hey, do I have any foo?"  "Where is my bar?"  And eventually, "Please re-order my baz."

**Comment by iyzpx** [comment]
Just a few thoughts:
- BOM Check (the projects function in Part-DB works perfect for production. But sometimes I just want to check a BOM without import the parts)
- Things like #274 could be realized with an API instead of implementing this in Part-DB
- Interacting with our inventory directly in KiCad
- We use https://github.com/openscopeproject/InteractiveHtmlBom for smb assembly. Since we have a lot of parts, a column with a storage location would be very handy
- Creating bots that search for parts marked as "Inform when available again somehwere"
- Creating bots for part health check (availability, manufacturing status)
- Push notifications if a stock is below minimum stock

I guess the most of us don't want to edit something. Mostly getters
searchPartByName, getPartByName, getPartById, getStorage, getAllParts? getAllStorageLocations?, getAllFootprints?
something like that

**Comment by SilentResonance** [comment]
My use cases for an API would be:
- Desktop app (like with Electron JS) to connect physical barcode readers (via COM/UART port)

> Example: Scan 2D datamatrix (e.g. digikey or mouser) on new arrived part bag. Then ask user what to do... add all to already available stock, create new part with information from the 2D code. Of course removal or change in stock by grabbing a bag, scan the code.
> All these changes could then be made with the API and the software could also be a systray application.

Or:

> Open the part (if it exists) and pre-fill quantity change fields or if part does not exist use ANSI MH10.8.2 data encoded to fill new part dialog automatically,

- Location movement system:

> Example: The circuit board assembler is using several bags with components for assembly. This leads to bags not being in normal storage location. Now if another developer needs parts (from storage or assembly) or needs to count and therefore find the bag, there might be a 'last seen' field for every part/every stock of a part (there might be multiple bags).
> In order to change this, every bag contains an RFID tag or extra barcode which is scanned with a reader terminal. These are located at each office or workshop entrance. Depending on where the RFID tag/barcode is scanned, the 'last seen' is changed (via API) or cleared if scanned at the storage terminal.
> 

These are just some ideas.... Electron JS might also be able to just open the part-db website internally and do the changes via normal GET/POST requests like the frontend would do.
Why Electron JS? Using COM/UART barcode scanners instead of keyboard wedge type. This leads to exclusive data input not put into the wrong focused field and therefore ease of use.
Electron JS seems to integrate with web services easily.

**Comment by jbtronics** [comment]
I am currently working on implementing the API. I am using [API Platform](https://api-platform.com/) for this, which allows for an easy implementation of an REST API (so using the HTTP GET, POST, DELETE and PATCH operations). The data is given and returned in JSON or preferably JSON+LD, which allows for a standardized way to link between different objects.

API Platforms automatically generates an OpenAPI/Swagger Schema, which documents the API in a standardized way and allows to generate a nice interactive API documentation. You can also use this OpenAPI schema to automatically generate clients to interact with Part-DB in various programming languages.

For authentication: A user can generate an API token on his settings page, which then the external application can then use to perform actions on the user behalf. I plan to have different access levels, so that you can for example restrict an API token to read-only access. The API tokens are always acting on behalf of a user (and is limited by its configured), so depending on the use case, you might want to create one (or more) separate users for API access.

**Comment by jbtronics** [comment]
There is now an API for Part-DB starting with Part-DB 1.8: See https://github.com/Part-DB/Part-DB-server/discussions/393

**Comment by niclash** [comment]
I am writing an client (Java) to the API at the moment. I have just get started, but basically I want to auto import my existing BOMs, which contains the LCSC numbers in them. So, I figured out the overall access to the API, and I can create/read/delete Manufacturers, Categories, Parts and so on. 
HOWEVER, I have not been able to figure out how (or even IF possible) to do the "Create parts from info provider" via the API. Looking at the web UI, one would assume that it is available on `/api/tools/info_providers/search`, but that gives an 404 error. And I think I have tried all variants on that.

So,
1. Is it supposed to work?
2. If so, how to access it?
3. If not supposed to work, point me in the right direction in the codebase and I'll see if I can make it work.

**Reply by jbtronics** [reply]
That is not implemented in the API (yet).

Only the endpoints in the API documentation, exist.
